File: src/main/java/com/example/dpop/id_fsc/FscCode.java
package com.example.dpop.id_fsc;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.Instant;

@Entity
@Table(name = "fsc_code")
public class FscCode {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long personId;

    private String code;

    private Instant expiresAt;

    protected FscCode() {
    }

    public FscCode(Long personId, String code, Instant expiresAt) {
        this.personId = personId;
        this.code = code;
        this.expiresAt = expiresAt;
    }

    public Long getId() {
        return id;
    }

    public Long getPersonId() {
        return personId;
    }

    public String getCode() {
        return code;
    }

    public Instant getExpiresAt() {
        return expiresAt;
    }

    public boolean isValid() {
        return expiresAt != null && Instant.now().isBefore(expiresAt);
    }
}

File: src/main/kotlin/com/example/dpop/id_fsc/internal/FscCode.kt
package com.example.dpop.id_fsc.internal

import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant

@Entity
@Table(name = "fsc_code")
class FscCode(
    var personId: Long? = null,

    /**
     * SHA-256 of the code, never the code itself. Unlike the six-digit TANs elsewhere this needs
     * no pepper: a Freischaltcode is not an enumerable number range, and a plain digest is what
     * lets the seeded rows be migrated in SQL (V15).
     */
    var codeHash: String? = null,

    var expiresAt: Instant? = null
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null

    /**
     * Expiry is the only limit: a Freischaltcode is fachlich reusable until it runs out. What
     * bounds GUESSING it is IdentThrottleService, not consumption.
     */
    val isValid: Boolean
        get() = expiresAt != null && Instant.now().isBefore(expiresAt)
}
